### PR TITLE
popover api

### PR DESCRIPTION
**Pull request by 12944qwerty**

### CODE REVIEW

File: src/types/coremods/message.ts
Comment (asportnoy) [inline src/types/coremods/message.ts:14]:
```suggestion
export type GetButtonItem = (message: Message, channel: Channel) => ButtonItem | null;
```

File: src/renderer/replugged.ts
Comment (asportnoy) [inline src/renderer/replugged.ts:30]: Import this above the namespace and export it directly in the namespace instead of assigning it after the fact

File: src/renderer/apis/message-popover.ts
Comment (asportnoy) [inline src/renderer/apis/message-popover.ts:16]: Instead of using an identifier based system, make `buttons` a `Set`. The `addButton` function can return a function that can be called to remove it.

**Review by asportnoy — CHANGES_REQUESTED** [review]
This should hook into the injector to make it easier to use. I'd imagine something like:
```ts
import { Injector } from 'replugged';

const injector = new Injector();
export function start() {
    innjector.utils.addPopoverButton(/* ... */);
}

export function stop() {
    injector.uninjectAll(); // Would also remove the popover button
}
```

File: src/renderer/replugged.ts
Comment (12944qwerty) [reply src/renderer/replugged.ts:30]: Lol I had this already, I just forgot to commit it

**Review by lexisother — CHANGES_REQUESTED** [review]
Blocking so I can review the plaintexts before it gets merged.

File: src/renderer/apis/message-popover.ts
Comment (asportnoy) [inline src/renderer/apis/message-popover.ts:19]: Add back the remove function

File: src/renderer/apis/message-popover.ts
Comment (asportnoy) [inline src/renderer/apis/message-popover.ts:41]: Can you add some more context here?

File: src/renderer/replugged.ts
Comment (asportnoy) [inline src/renderer/replugged.ts:29]: I don't think you need the type here anymore, it should be able to infer that.

**Review by asportnoy — CHANGES_REQUESTED** [review]

File: src/renderer/apis/message-popover.ts
Comment (asportnoy) [inline src/renderer/apis/message-popover.ts:51]:
```suggestion
          logger.error(`Error in making the button [${item?.key}]`, err, item);
```

File: src/renderer/apis/message-popover.ts
Comment (asportnoy) [inline src/renderer/apis/message-popover.ts:54]:
```suggestion
        logger.error("Error while running GetButtonItem function", err, getItem);
```

File: src/renderer/modules/injector.ts
Comment (asportnoy) [inline src/renderer/modules/injector.ts:280]: Add an example here

File: src/renderer/modules/injector.ts
Comment (asportnoy) [inline src/renderer/modules/injector.ts:283]: Import this function instead of calling it on the window

**Review by asportnoy — CHANGES_REQUESTED** [review]

File: src/renderer/apis/message-popover.ts
Comment (asportnoy) [inline src/renderer/apis/message-popover.ts:18]: Nit
```suggestion
    return () => this.removeButton(item);
```

File: src/renderer/modules/injector.ts
Comment (12944qwerty) [reply src/renderer/modules/injector.ts:283]: The class has to be initialized... which it is in replugged.ts. Can't import it from there...

Comment (asportnoy) [reply src/renderer/modules/injector.ts:283]: Switch it to a namespace instead of class. That way it doesn't need to be initialized at all.

You could also initialize it within that file, but a namespace is much better for this.

File: src/renderer/modules/injector.ts
Comment (asportnoy) [inline src/renderer/modules/injector.ts:285]: Include importing code like with https://docs.replugged.dev/classes/injector-1.Injector.html

File: src/renderer/modules/injector.ts
Comment (asportnoy) [inline src/renderer/modules/injector.ts:294]: Can you be more clear on what this does?

**Review by asportnoy — CHANGES_REQUESTED** [review]

**Comment by colin273** [comment]
Possibly a stupid question, but is the key actually used anywhere?

**Comment by asportnoy** [comment]
It's used by React

File: src/renderer/coremods/messagePopover/plaintextPatches.ts
Comment (lexisother) [inline src/renderer/coremods/messagePopover/plaintextPatches.ts:5]:
```suggestion
  {
    find: 'navId:"message-actions",',
    replacements: [
```

**Review by lexisother — COMMENTED** [review]
May help reduce the load of the operation.

File: src/renderer/apis/message-popover.ts
Comment (asportnoy) [inline src/renderer/apis/message-popover.ts]: Generate the key here, not below

**Review by asportnoy — CHANGES_REQUESTED** [review]

File: src/renderer/apis/message-popover.ts
Comment (asportnoy) [inline src/renderer/apis/message-popover.ts:84]: Debug crust
```suggestion
```

**Review by lexisother — APPROVED** [review]

**Review by asportnoy — APPROVED** [review]

File: src/renderer/coremods/messagePopover/plaintextPatches.ts
Comment (asportnoy) [inline src/renderer/coremods/messagePopover/plaintextPatches.ts:9]: Doesn't work anymore, probably a Discord change

**Review by asportnoy — CHANGES_REQUESTED** [review]

**Review by asportnoy — APPROVED** [review]
Works

**Review by colin273 — APPROVED** [review]

File: src/renderer/modules/common/constants.ts
Comment (asportnoy) [inline src/renderer/modules/common/constants.ts:5]: Revert the changes you made here

File: cspell.json
Comment (asportnoy) [inline cspell.json:47]: Just camel case the variable and you won't need this

**Review by asportnoy — CHANGES_REQUESTED** [review]

File: src/renderer/modules/common/constants.ts
Comment (12944qwerty) [reply src/renderer/modules/common/constants.ts:5]: This was lint....

Comment (asportnoy) [reply src/renderer/modules/common/constants.ts:5]: It should not have done that

**Review by asportnoy — APPROVED** [review]

**Comment by SammCheese** [comment]
Just merge this shid already. Ya'll worse than ven when it comes to merging delays.

**Comment by colin273** [comment]
@SammCheese Nop

Jk, I think the API is ready to merge so I'll approve it now, but I've been intentionally blocking this until I'm (1) satisfied with the sanity of the API and (2) available to think clearly about API design. There's always a method to the madness, and that method is quality control.

**Review by colin273 — APPROVED** [review]

**Pull request merged**
